pkg/api/steve/collections/rbac.go

```diff
@@ -0,0 +1,315 @@
+			if err != nil {
+				return nil, err
+			}
+			return rb, nil
+		}
+		if _, ok := collectionClusterRole.GetAnnotations()[clusterRoleOwnerAnnotation]; ok { // owned by roletemplate controllers, don't update it
+			return rb, nil
+		}
+		err = h.compareAndUpdateClusterRole(collectionClusterRole, collectionRules, clusterRole)
+		if err != nil {
+			return nil, err
+		}
+		return rb, nil
+	}
+	collectionRole, err := h.rolesCache.Get(namespace.Name, collectionRoleBinding.RoleRef.Name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+	if collectionRole == nil {
+		err = h.createRole(collectionRoleBinding.RoleRef.Name, namespace.Name, collectionRules, role)
+		if err != nil {
+			return nil, err
+		}
+		return rb, nil
+	}
+	err = h.compareAndUpdateRole(collectionRole, collectionRules, role)
+	if err != nil {
+		return nil, err
+	}
+
+	return rb, nil
+}
+
+func GetCollectionRules(apis APIResourceWatcher, rules []rbacv1.PolicyRule) ([]rbacv1.PolicyRule, error) {
+	collectionRules := []rbacv1.PolicyRule{}
+	validRules := []rbacv1.PolicyRule{}
+	for _, rule := range rules {
+		if len(rule.ResourceNames) != 0 {
+			continue
+		}
+		if len(rule.NonResourceURLs) != 0 {
+			continue
+		}
+		for _, verb := range rule.Verbs {
+			if verb == "*" || verb == "list" {
+				validRules = append(validRules, rule)
+			}
+		}
+	}
+	if len(validRules) == 0 {
+		return collectionRules, nil
+	}
+	if len(apis.List()) == 0 {
+		return collectionRules, ErrNotReady
+	}
+	for _, rule := range validRules {
+		for _, apiGroup := range rule.APIGroups {
+			for _, resource := range rule.Resources {
+				_, ok := apis.Get(resource, apiGroup)
+				if ok {
+					if len(collectionRules) == 0 {
+						collectionRules = []rbacv1.PolicyRule{
+							{
+								APIGroups: []string{Group},
+								Verbs:     []string{"list"},
+							},
+						}
+					}
+					if apiGroup != "" {
+						resource = apiGroup + "." + resource
+					}
+					collectionRules[0].Resources = append(collectionRules[0].Resources, resource)
+					continue
+				}
+				if resource == "*" {
+					if apiGroup == "*" {
+						collectionRules = []rbacv1.PolicyRule{
+							{
+								APIGroups: []string{Group},
+								Verbs:     []string{"list"},
+								Resources: []string{"*"},
+							},
+						}
+						return collectionRules, nil
+					}
+					apis := apis.List()
+					for _, api := range apis {
+						if api.Group == apiGroup {
+							if len(collectionRules) == 0 {
+								collectionRules = []rbacv1.PolicyRule{
+									{
+										APIGroups: []string{Group},
+										Verbs:     []string{"list"},
+									},
+								}
+							}
+							collectionRules[0].Resources = append(collectionRules[0].Resources, api.Name)
+						}
+					}
+					if len(collectionRules) > 0 {
+						sort.Strings(collectionRules[0].Resources)
+					}
+					return collectionRules, nil
+				}
+				if apiGroup == "*" {
+					apis := apis.List()
```

Comment:
doing the check for the apiGroup == "*" in the inmost loop could lead to you re-calculating this multiple times. I would recommend checking in an outer loop instead. something like:

```go
for _, rule := range rules{
    var apiGroups []string
    var isAllGroups bool
    // if we have a "*", fan out the groups and make sure the resources know allGroups are to be considered
    for _, group := range rule.apiGroups{
        if group == "*"{
            isAllGroups = true
            apiGroups = apis.List()
            break
        }
        apiGroups = append(apiGroups, group)
    }
    for _, resource := range resources {
        // use already calculated groups here
    }
}
```

Reply:
Not sure if I'm understanding your example correctly. I think the inefficiency is built in to the way policy rules are structured, which is that there isn't a 1:1 mapping from resource to apigroup and you have to cross-match them all. So if I have

```
rules:
- apiGroups:
  - "networking.k8s.io"
  - "storage.k8s.io"
  resources:
  - "ingresses"
  - "storageclasses"
```
I have to check discovery for "networking.k8s.io.ingresses", "networking.k8s.io.storageclasses", "storage.k8s.io.ingresses" and "storage.k8s.io.storageclasses". So assuming none of the apigroups is "*", in the resources loop I still have to loop over the apigroups list again to try to find a resource/apigroup combination that exists.

Reply:
Resolving, approach changed, file no longer exists.